README.md
# Proby
A simple library for working with the Proby task monitoring application.

[![Build Status](https://secure.travis-ci.org/signal/proby-ruby.png?branch=master)](http://travis-ci.org/signal/proby-ruby)

Installation
------------

### RubyGems ###
Proby can be installed using RubyGems

    gem install proby

Inside your script, be sure to

    require "rubygems"
    require "proby"

### Bundler ###
If you're using Bundler, add the following to your Gemfile

    gem "proby"

and then run

    bundle install


Setup
-----
Before notifications can be sent, you must tell Proby your API key.  This only needs to be done once,
and should ideally be done inside your apps initialization code.

    Proby.api_key = "b4fe1200c105012efde3482a1411a947"

In addition, you can optionally give Proby a logger to use.

    Proby.logger = Rails.logger


Sending Notifications
---------------------
The easiest way to have Proby monitor your task is by wrapping your code in a call to Proby's `monitor` function.

    Proby.monitor(task_api_id) do
      # Do something here
    end

You can also send the start and finish notifications manually via calls to `send_start_notification` and `send_finish_notification`.

    Proby.send_start_notification(task_api_id)
    # Do something here
    Proby.send_finish_notification(task_api_id)

Specifying the `task_api_id` when calling any of the the notification methods is optional.  If it is not provided,
Proby will use the value of the `PROBY_TASK_ID` environment variable.  If no task id is specified
in the method call, and no value is set in the `PROBY_TASK_ID` environment variable, then no notification
will be sent.


The Resque Plugin
-----------------
The Resque plugin will automatically send start and finish notifications to Proby when your job
starts and finishes.  Simply `extend Proby::ResquePlugin` in your Resque job.  The task id
can either be pulled from the `PROBY_TASK_ID` environment variable, or specified in the job itself
by setting the `@proby_id` attribute to the task id.

    class SomeJob
      extend Proby::ResquePlugin
      @proby_id = 'abc123'  # Or simply let it use the value in the PROBY_TASK_ID environment variable

      self.perform
        do_stuff
      end
    end


Managing Tasks
--------------
The Proby::ProbyTask class can be used to create, read, update, delete, pause, and unpause your
tasks on Proby.

    my_tasks = Proby::ProbyTask.find(:all)
    a_specific_task = Proby::ProbyTask.find("the_proby_task_id")

    task = Proby::ProbyTask.create(:name => 'Task name', :crontab => '* * * * *')

    task.name = "New name"
    task.save

    task.pause
    task.unpause

    task.delete


API Doc
-------
[http://rdoc.info/github/signal/proby-ruby/master/frames](http://rdoc.info/github/signal/proby-ruby/master/frames)



HISTORY.md
2.2.0
-----
* Loosened up the version dependency on multi_json
* Log error to stderr when API key is not set
* Fix warning about version already being defined

2.1.0
-----
* Added Proby.monitor method

2.0.0
-----
* Initial import from proby_notifier codebase.
* Re-wrote core internals.
* Added Proby::ProbyTask to wrap the Task API.


test/test_helper.rb
require 'rubygems'
require 'bundler'
begin
  Bundler.setup(:default, :development)
rescue Bundler::BundlerError => e
  $stderr.puts e.message
  $stderr.puts "Run `bundle install` to install missing gems"
  exit e.status_code
end

require 'test/unit'
require 'fakeweb'
require 'shoulda'
require 'mocha'

require 'proby'



test/notifier_test.rb
require 'test_helper'

class NotifierTest < Test::Unit::TestCase

  def setup
    ENV['PROBY_TASK_ID'] = nil
    FakeWeb.clean_registry
    FakeWeb.allow_net_connect = false
  end

  def teardown
    FakeWeb.allow_net_connect = true
  end

  should "not send the notification if the api_key is not set" do
    previous_stderr, $stderr = $stderr, StringIO.new
    begin
      assert_nil Proby.send_start_notification("abc123")
    ensure
      $stderr = previous_stderr
    end
  end

  context "with an api key set" do
    setup do
      Proby.api_key = '1234567890abcdefg'
    end

    should "not send the notification if a task id was not specified" do
      assert_nil Proby.send_start_notification
    end

    should "not send the notification if a task id is blank" do
      assert_nil Proby.send_start_notification("  ")
    end

    should "send a start notification if a task_id is specified in the call" do
      FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + "/api/v1/tasks/abc123xyz456/start.json", :status => ["200", "OK"])
      assert_equal 200, Proby.send_start_notification("abc123xyz456")
    end

    should "send a start notification if a task_id is specified in an environment variable" do
      ENV['PROBY_TASK_ID'] = "uuu777sss999"
      FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + "/api/v1/tasks/uuu777sss999/start.json", :status => ["200", "OK"])
      assert_equal 200, Proby.send_start_notification
    end

    should "send a finish notification if a task_id is specified in the call" do
      FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + "/api/v1/tasks/abc123xyz456/finish.json", :status => ["200", "OK"])
      assert_equal 200, Proby.send_finish_notification("abc123xyz456")
    end

    should "send a finish notification with options if options are specified" do
      FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + "/api/v1/tasks/abc123xyz456/finish.json", :status => ["200", "OK"])
      assert_equal 200, Proby.send_finish_notification("abc123xyz456", :failed => true, :error_message => "something bad happened")
    end

    should "send a finish notification if a task_id is specified in an environment variable" do
      ENV['PROBY_TASK_ID'] = "iii999ooo222"
      FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + "/api/v1/tasks/iii999ooo222/finish.json", :status => ["200", "OK"])
      assert_equal 200, Proby.send_finish_notification
    end

    should "send a start and finish notificaiton when using the monitor method" do
      Proby.expects(:send_start_notification).with("abc123xyz456")
      Proby.expects(:send_finish_notification).with() { |param1, param2| param1 == "abc123xyz456" && param2[:failed] == false && param2[:error_message].nil? }
      assert_equal "foo", Proby.monitor("abc123xyz456") { "foo" }
    end

    should "include error information in the finish notification if the block passed to monitor raises an exception" do
      Proby.expects(:send_start_notification).with("abc123xyz456")
      Proby.expects(:send_finish_notification).with() { |param1, param2| param1 == "abc123xyz456" && param2[:failed] == true && param2[:error_message].include?("This is the error message") }
      e = assert_raise Exception do
        Proby.monitor("abc123xyz456") { raise Exception.new("This is the error message") }
      end
      assert_equal "This is the error message", e.message
    end
  end

end


test/proby_task_test.rb
require 'test_helper'

class ProbyTaskTest < Test::Unit::TestCase

  def setup
    Proby.api_key = '1234567890abcdefg'
    FakeWeb.clean_registry
    FakeWeb.allow_net_connect = false
  end

  def teardown
    FakeWeb.allow_net_connect = true
  end

  should "raise an error if the api key has not been set" do
    Proby.api_key = nil
    e = assert_raise(Proby::InvalidApiKeyException) do
      Proby::ProbyTask.find(:all)
    end
    assert_equal "Your Proby API key has not been set.  Set it using Proby.api_key = 'my_api_key'", e.message
  end

  should "be able to get a list of tasks" do
    FakeWeb.register_uri(:get, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks.json', :status => ['200', 'OK'], :body => <<-END)
{
  "tasks": [
    {
      "name": "task 1",
      "api_id": "abc123",
      "crontab": "* * * * *",
      "paused": false,
      "time_zone": "UTC",
      "machine": null,
      "finish_alarms_enabled": true,
      "maximum_run_time": null,
      "start_notification_grace_period": 5,
      "consecutive_alarmed_tasks": 0,
      "consecutive_alarmed_tasks_required_to_trigger_alarm": 1,
      "created_at": "2012-04-25 20:10:20 UTC",
      "updated_at": "2012-04-25 20:10:20 UTC",
      "status": {
        "description": "OK",
        "details": ""
      }
    },
    {
      "name": "task 2",
      "api_id": "abc456",
      "crontab": "* * * * *",
      "paused": false,
      "time_zone": "UTC",
      "machine": "dopey",
      "finish_alarms_enabled": true,
      "maximum_run_time": 90,
      "start_notification_grace_period": 5,
      "consecutive_alarmed_tasks": 3,
      "consecutive_alarmed_tasks_required_to_trigger_alarm": 1,
      "created_at": "2012-04-25 20:10:20 UTC",
      "updated_at": "2012-04-25 20:10:20 UTC",
      "status": {
        "description": "OK",
        "details": "Wonderful"
      }
    },
    {
      "name": "task 3",
      "api_id": "abc789",
      "crontab": "1 2 3 4 5",
      "paused": false,
      "time_zone": "UTC",
      "machine": null,
      "finish_alarms_enabled": true,
      "maximum_run_time": null,
      "start_notification_grace_period": 5,
      "consecutive_alarmed_tasks": 0,
      "consecutive_alarmed_tasks_required_to_trigger_alarm": 1,
      "created_at": "2012-04-25 20:10:20 UTC",
      "updated_at": "2012-04-25 20:10:20 UTC",
      "status": {
        "description": "OK",
        "details": ""
      }
    }
  ]
}
END
    tasks = Proby::ProbyTask.find(:all)
    assert_equal 3, tasks.size

    task = tasks[1]
    assert_equal "task 2", task.name
    assert_equal "abc456", task.api_id
    assert_equal "* * * * *", task.crontab
    assert_equal false, task.paused
    assert_equal "UTC", task.time_zone
    assert_equal "dopey", task.machine
    assert_equal true, task.finish_alarms_enabled
    assert_equal 90, task.maximum_run_time
    assert_equal 5, task.start_notification_grace_period
    assert_equal 3, task.consecutive_alarmed_tasks
    assert_equal 1, task.consecutive_alarmed_tasks_required_to_trigger_alarm
    assert_equal Time.utc(2012, 4, 25, 20, 10, 20), task.created_at
    assert_equal Time.utc(2012, 4, 25, 20, 10, 20), task.updated_at
    assert_equal "OK", task.status.description
    assert_equal "Wonderful", task.status.details
  end

  should "raise an exception if unable to get a list of tasks" do
    FakeWeb.register_uri(:get, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks.json', :status => ['400', 'Bad request'], :body => <<-END)
{
  "request" : "http://www.example.com/api/v1/tasks.json",
  "message" : "Something bad happened"
}
END
    e = assert_raise Proby::ApiException do
      Proby::ProbyTask.find(:all)
    end
    assert e.message.include?("API request failed with a response code of 400")
    assert e.message.include?("Something bad happened")
  end

  should "be able to display a specific task" do
    FakeWeb.register_uri(:get, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123.json', :status => ['200', 'OK'], :body => <<-END)
{
  "task": {
    "name": "task 1",
    "api_id": "abc123",
    "crontab": "* * * * *",
    "paused": false,
    "time_zone": "UTC",
    "machine": "dopey",
    "finish_alarms_enabled": true,
    "maximum_run_time": 90,
    "start_notification_grace_period": 5,
    "consecutive_alarmed_tasks": 3,
    "consecutive_alarmed_tasks_required_to_trigger_alarm": 1,
    "created_at": "2012-04-25 20:10:20 UTC",
    "updated_at": "2012-04-25 20:10:20 UTC",
    "status": {
      "description": "OK",
      "details": "Wonderful"
    }
  }
}
END

    task = Proby::ProbyTask.find("abc123")
    assert_equal "task 1", task.name
    assert_equal "abc123", task.api_id
    assert_equal "* * * * *", task.crontab
    assert_equal false, task.paused
    assert_equal "UTC", task.time_zone
    assert_equal "dopey", task.machine
    assert_equal true, task.finish_alarms_enabled
    assert_equal 90, task.maximum_run_time
    assert_equal 5, task.start_notification_grace_period
    assert_equal 3, task.consecutive_alarmed_tasks
    assert_equal 1, task.consecutive_alarmed_tasks_required_to_trigger_alarm
    assert_equal Time.utc(2012, 4, 25, 20, 10, 20), task.created_at
    assert_equal Time.utc(2012, 4, 25, 20, 10, 20), task.updated_at
    assert_equal "OK", task.status.description
    assert_equal "Wonderful", task.status.details
  end

  should "return nil if the specific task could not be found" do
    FakeWeb.register_uri(:get, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123.json', :status => ['404', 'Not Found'])
    assert_nil Proby::ProbyTask.find("abc123")
  end

  should "raise an exception if unable to fetch a specific task" do
    FakeWeb.register_uri(:get, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123.json', :status => ['400', 'Bad request'], :body => <<-END)
{
  "request" : "http://www.example.com/api/v1/tasks/abc123.json",
  "message" : "Something bad happened"
}
END
    e = assert_raise Proby::ApiException do
      Proby::ProbyTask.find("abc123")
    end
    assert e.message.include?("API request failed with a response code of 400")
    assert e.message.include?("Something bad happened")
  end

  should "raise an error if trying to fetch a task without specifying the api_id" do
    assert_raise(Proby::InvalidParameterException) { Proby::ProbyTask.find(nil) }
    assert_raise(Proby::InvalidParameterException) { Proby::ProbyTask.find(" ") }
  end

  should "be able to create a new task" do
    FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks.json', :status => ['201', 'Created'], :body => <<-END)
{
  "task": {
    "name": "task 1",
    "api_id": "abc123",
    "crontab": "* * * * *",
    "paused": false,
    "time_zone": "UTC",
    "machine": "dopey",
    "finish_alarms_enabled": true,
    "maximum_run_time": 90,
    "start_notification_grace_period": 5,
    "consecutive_alarmed_tasks": 3,
    "consecutive_alarmed_tasks_required_to_trigger_alarm": 1,
    "created_at": "2012-04-25 20:10:20 UTC",
    "updated_at": "2012-04-25 20:10:20 UTC",
    "status": {
      "description": "OK",
      "details": "Wonderful"
    }
  }
}
END

    # Not specifying all attributes, for berivity
    task = Proby::ProbyTask.create(:name => "abc123", :crontab => "* * * * *")

    assert_equal "task 1", task.name
    assert_equal "abc123", task.api_id
    assert_equal "* * * * *", task.crontab
    assert_equal false, task.paused
    assert_equal "UTC", task.time_zone
    assert_equal "dopey", task.machine
    assert_equal true, task.finish_alarms_enabled
    assert_equal 90, task.maximum_run_time
    assert_equal 5, task.start_notification_grace_period
    assert_equal 3, task.consecutive_alarmed_tasks
    assert_equal 1, task.consecutive_alarmed_tasks_required_to_trigger_alarm
    assert_equal Time.utc(2012, 4, 25, 20, 10, 20), task.created_at
    assert_equal Time.utc(2012, 4, 25, 20, 10, 20), task.updated_at
    assert_equal "OK", task.status.description
    assert_equal "Wonderful", task.status.details
  end

  should "raise an exception if unable to create a task" do
    FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks.json', :status => ['400', 'Bad request'], :body => <<-END)
{
  "request" : "http://www.example.com/api/v1/tasks.json",
  "message" : "Something bad happened"
}
END
    e = assert_raise Proby::ApiException do
      Proby::ProbyTask.create(:name => "foo", :crontab => "invalid")
    end
    assert e.message.include?("API request failed with a response code of 400")
    assert e.message.include?("Something bad happened")
  end

  should "raise an error if trying to create a task without specifying the attributes" do
    assert_raise(Proby::InvalidParameterException) { Proby::ProbyTask.create(nil) }
    assert_raise(Proby::InvalidParameterException) { Proby::ProbyTask.create({}) }
  end

  should "raise an error if trying to create a task without the necessary required attributes" do
    assert_raise(Proby::InvalidParameterException) { Proby::ProbyTask.create(:name => "Foo") }
    assert_raise(Proby::InvalidParameterException) { Proby::ProbyTask.create(:crontab => "* * * * *") }
  end

  should "be able to update a task" do
    FakeWeb.register_uri(:put, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123.json', :status => ['200', 'OK'])
    proby_task = Proby::ProbyTask.new('api_id' => 'abc123', 'name' => 'Test task', 'crontab' => '* * * * *')
    proby_task.name = "New name"
    proby_task.crontab = "1 2 3 4 5"
    proby_task.time_zone = "UTC"
    proby_task.machine = "sleepy"
    proby_task.finish_alarms_enabled = "false"
    proby_task.maximum_run_time = 60
    proby_task.start_notification_grace_period = 10
    proby_task.consecutive_alarmed_tasks_required_to_trigger_alarm = 4
    assert proby_task.save
  end

  should "raise an exception if unable to update a task" do
    FakeWeb.register_uri(:put, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123.json', :status => ['400', 'Bad request'], :body => <<-END)
{
  "request" : "http://www.example.com/api/v1/tasks/abc123.json",
  "message" : "Something bad happened"
}
END
    proby_task = Proby::ProbyTask.new('api_id' => 'abc123', 'name' => 'Test task', 'crontab' => '* * * * *')
    proby_task.name = "New name"
    proby_task.crontab = "1 2 3 4 5"
    e = assert_raise Proby::ApiException do
      proby_task.save
    end
    assert e.message.include?("API request failed with a response code of 400")
    assert e.message.include?("Something bad happened")
  end

  should "raise an error if trying to update a task without the necessary required attributes" do
    proby_task = Proby::ProbyTask.new('api_id' => 'abc123', 'name' => 'Test task', 'crontab' => '* * * * *')

    proby_task.name = nil
    assert_raise(Proby::InvalidParameterException) { proby_task.save }

    proby_task.name = "  "
    assert_raise(Proby::InvalidParameterException) { proby_task.save }

    proby_task.crontab = nil
    assert_raise(Proby::InvalidParameterException) { proby_task.save }

    proby_task.crontab = "  "
    assert_raise(Proby::InvalidParameterException) { proby_task.save }
  end

  should "be able to delete a task" do
    FakeWeb.register_uri(:delete, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123.json', :status => ['200', 'OK'])
    proby_task = Proby::ProbyTask.new('api_id' => 'abc123', 'name' => 'Test task', 'crontab' => '* * * * *')
    assert proby_task.delete
    assert proby_task.frozen?
    begin
      proby_task.name = "foo"
      assert false, "Should have raised an exception"
    rescue Exception => e
      assert e.message.include?("can't modify frozen")
    end
  end

  should "raise an exception if unable to delete a task" do
    FakeWeb.register_uri(:delete, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123.json', :status => ['400', 'Bad request'], :body => <<-END)
{
  "request" : "http://www.example.com/api/v1/tasks/abc123.json",
  "message" : "Something bad happened"
}
END
    proby_task = Proby::ProbyTask.new('api_id' => 'abc123', 'name' => 'Test task', 'crontab' => '* * * * *')
    e = assert_raise Proby::ApiException do
      proby_task.delete
    end
    assert e.message.include?("API request failed with a response code of 400")
    assert e.message.include?("Something bad happened")
  end

  should "be able to pause a task" do
    FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123/pause.json', :status => ['200', 'OK'])
    proby_task = Proby::ProbyTask.new('api_id' => 'abc123', 'name' => 'Test task', 'crontab' => '* * * * *', 'paused' => 'false')
    assert proby_task.pause
    assert_equal true, proby_task.paused
  end

  should "raise an exception if unable to pause a task" do
    FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123/pause.json', :status => ['400', 'Bad request'], :body => <<-END)
{
  "request" : "http://www.example.com/api/v1/tasks/abc123/pause.json",
  "message" : "Something bad happened"
}
END
    proby_task = Proby::ProbyTask.new('api_id' => 'abc123', 'name' => 'Test task', 'crontab' => '* * * * *')
    e = assert_raise Proby::ApiException do
      proby_task.pause
    end
    assert e.message.include?("API request failed with a response code of 400")
    assert e.message.include?("Something bad happened")
  end

  should "be able to unpause a task" do
    FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123/unpause.json', :status => ['200', 'OK'])
    proby_task = Proby::ProbyTask.new('api_id' => 'abc123', 'name' => 'Test task', 'crontab' => '* * * * *', 'paused' => 'true')
    assert proby_task.unpause
    assert_equal false, proby_task.paused
  end

  should "raise an exception if unable to unpause a task" do
    FakeWeb.register_uri(:post, Proby::ProbyHttpApi.base_uri + '/api/v1/tasks/abc123/unpause.json', :status => ['400', 'Bad request'], :body => <<-END)
{
  "request" : "http://www.example.com/api/v1/tasks/abc123/unpause.json",
  "message" : "Something bad happened"
}
END
    proby_task = Proby::ProbyTask.new('api_id' => 'abc123', 'name' => 'Test task', 'crontab' => '* * * * *')
    e = assert_raise Proby::ApiException do
      proby_task.unpause
    end
    assert e.message.include?("API request failed with a response code of 400")
    assert e.message.include?("Something bad happened")
  end

end


lib/proby.rb
require 'logger'
require 'httparty'
require 'chronic'

require 'proby/exceptions'
require 'proby/proby_http_api'
require 'proby/proby_task'
require 'proby/notifier'
require 'proby/resque_plugin'

module Proby

  # A simple library for working with the Proby task monitoring application.
  class << self

    # Set your Proby API key.
    #
    # @param [String] api_key Your Proby API key
    #
    # @example
    #   Proby.api_key = '1234567890abcdefg'
    def api_key=(api_key)
      @api_key = api_key
    end

    # Get the api key.
    def api_key
      @api_key
    end

    # Set the logger to be used by Proby.
    #
    # @param [Logger] logger The logger you would like Proby to use
    #
    # @example
    #   Proby.logger = Rails.logger
    #   Proby.logger = Logger.new(STDERR)
    def logger=(logger)
      @logger = logger
    end

    # Get the logger used by Proby.
    def logger
      @logger ||= Logger.new("/dev/null")
    end

    # Send a start notification for this task to Proby.
    #
    # @param [String] proby_task_id The id of the task to be notified. If nil, the
    #                               value of the +PROBY_TASK_ID+ environment variable will be used.
    #
    # @return [Fixnum] The HTTP status code that was returned from Proby.
    def send_start_notification(proby_task_id=nil)
      Notifier.send_notification('start', proby_task_id)
    end

    # Send a finish notification for this task to Proby
    #
    # @param [String] proby_task_id The id of the task to be notified. If nil, the
    #                               value of the +PROBY_TASK_ID+ environment variable will be used.
    # @param [Hash] options The options for the finish notification
    # @option options [Boolean] :failed true if this task run resulted in some sort of failure. Setting
    #                                   this parameter to true will trigger a notification to be sent to
    #                                   the alarms configured for the given task. Defaults to false.
    # @option options [String] :error_message A string message describing the failure that occurred.
    #                                         1,000 character limit.
    #
    # @return [Fixnum] The HTTP status code that was returned from Proby.
    def send_finish_notification(proby_task_id=nil, options={})
      Notifier.send_notification('finish', proby_task_id, options)
    end

    # Surround the block of code with Proby start and finish notifications.  If an exception
    # is raised in the block of code, then the task will be marked as failed, and the
    # exception's message and backtrace will be sent to Proby as the task's error message.
    #
    # @param [String] proby_task_id The id of the task to be notified. If nil, the
    #                               value of the +PROBY_TASK_ID+ environment variable will be used.
    def monitor(proby_task_id=nil)
      failed = false
      error_message = nil
      begin
        Proby.send_start_notification(proby_task_id)
        yield
      rescue Exception => e
        failed = true
        error_message = "#{e.class.name}: #{e.message}"
        error_message << "\n#{e.backtrace.join("\n")}" if e.backtrace
        raise e
      ensure
        Proby.send_finish_notification(proby_task_id, :failed => failed, :error_message => error_message)
      end
    end

  end
end



lib/proby/proby_task.rb
module Proby

  # Represents the status of a Proby task
  class ProbyTaskStatus
    # The description of the task (OK, ERROR, PAUSED)
    attr_reader :description

    # Any details (why the task is in the ERROR state)
    attr_reader :details

    def initialize(attributes={})
      @description = attributes['description']
      @details = attributes['details']
    end
  end

  # Represents a task in Proby
  class ProbyTask < ProbyHttpApi
    # The name of the task
    attr_accessor :name

    # The schedule for the task, specified in crontab format
    attr_accessor :crontab

    # The time zone of the machine executing the task
    attr_accessor :time_zone

    # The name of the machine that is responsible for running this task
    attr_accessor :machine

    # Should finish alarms be sent when the task runs longer than expected?
    attr_accessor :finish_alarms_enabled

    # The maximum amount of time the task is allowed to run before Proby sends
    # a finish alarm. If not specified, Proby will determine when an alarm should
    # be sent based on past run times
    attr_accessor :maximum_run_time

    # The number of minutes to wait for a task to send its start notification after it
    # should have started before sending an alarm
    attr_accessor :start_notification_grace_period

    # The number of consecutive tasks that must fail before an alarm is sent
    attr_accessor :consecutive_alarmed_tasks_required_to_trigger_alarm

    # The API Task ID of the task
    attr_reader :api_id

    # Is the task currently paused?
    attr_reader :paused

    # The number of consecutive times this task has triggered an alarm
    attr_reader :consecutive_alarmed_tasks

    # The date and time this task was created
    attr_reader :created_at

    # The date and time this task was updated
    attr_reader :updated_at

    # The current status of the task, represented as a ProbyTaskStatus
    attr_reader :status

    # <b>Should not be called directly</b>
    def initialize(attributes={})
      @name = attributes['name']
      @api_id = attributes['api_id']
      @crontab = attributes['crontab']
      @paused = attributes['paused']
      @time_zone = attributes['time_zone']
      @machine = attributes['machine']
      @finish_alarms_enabled = attributes['finish_alarms_enabled']
      @maximum_run_time = attributes['maximum_run_time']
      @start_notification_grace_period = attributes['start_notification_grace_period']
      @consecutive_alarmed_tasks = attributes['consecutive_alarmed_tasks']
      @consecutive_alarmed_tasks_required_to_trigger_alarm = attributes['consecutive_alarmed_tasks_required_to_trigger_alarm']
      @created_at = Chronic.parse attributes['created_at']
      @updated_at = Chronic.parse attributes['updated_at']
      @status = ProbyTaskStatus.new(attributes['status']) if attributes['status']
    end

    # Get a single, or all tasks from Proby.
    #
    # @param [Object] param :all if you are fetching all tasks, or the api_id of the Proby task you would like to fetch.
    #
    # @return If requesting all tasks, an [Array<ProbyTask>] will be returned.  If an api_id was provided, the
    #           ProbyTask with that api_id will be returned if it exists, or nil if it could not be found.
    #
    # @example
    #   all_of_my_tasks = ProbyTask.find(:all)
    #   my_task = ProbyTask.find('my_proby_task_api_id')
    def self.find(param)
      ensure_api_key_set
      param == :all ? list : fetch(param)
    end

    # Create a new Proby task.
    #
    # @param [Hash] attributes The attributes for your task.
    # @option attributes [String] :name A name for your task.
    # @option attributes [String] :crontab The schedule of the task, specified in cron format.
    # @option attributes [String] :time_zone <b>(Optional)</b> The time zone of the machine executing the task.
    # @option attributes [String] :machine <b>(Optional)</b> The name of the machine that is responsible for running this task.
    #                               Will default to the default time zone configured in Proby if not specified.
    # @option attributes [Boolean] :finish_alarms_enabled <b>(Optional)</b> true if you would like to receive finish alarms for
    #                                this task, false otherwise (default: true).
    # @option attributes [Fixnum] :maximum_run_time <b>(Optional)</b> The maximum amount of time the task is allowed to run before
    #                               Proby sends a finish alarm. If not specified, Proby will determine when an alarm should be
    #                               sent based on past run times.
    # @option attributes [Fixnum] :start_notification_grace_period <b>(Optional)</b> The number of minutes to wait for a task to
    #                               send its start notification after it should have started before sending an alarm.
    # @option attributes [Fixnum] :consecutive_alarmed_tasks_required_to_trigger_alarm <b>(Optional)</b> The number of consecutive
    #                               tasks that must fail before an alarm is sent.
    #
    # @return [ProbyTask] The task that was created.
    #
    # @example
    #   proby_task = ProbyTask.create(:name => "My new task", :crontab => "* * * * *")
    def self.create(attributes={})
      ensure_api_key_set
      raise InvalidParameterException.new("attributes are required") if attributes.nil? || attributes.empty?
      raise InvalidParameterException.new("name is required") unless !blank?(attributes[:name]) || !blank?(attributes['name'])
      raise InvalidParameterException.new("crontab is required") unless !blank?(attributes[:crontab]) || !blank?(attributes['crontab'])

      Proby.logger.info "Creating task with attributes: #{attributes.inspect}"
      response = post("/api/v1/tasks.json",
                      :format => :json,
                      :body => MultiJson.encode(:task => attributes),
                      :headers => default_headers)

      if response.code == 201 
        new(response.parsed_response['task'])
      else
        handle_api_failure(response)
      end 
    end

    # Saves the task in Proby, updating all attributes to the values stored in the object.  Only the attributes specified in
    # the ProbyTask.create documentation can be updated.
    #
    # @example
    #   proby_task = ProbyTask.get('my_proby_task_api_id')
    #   proby_task.name = "Some other name"
    #   proby_task.crontab = "1 2 3 4 5"
    #   proby_task.save
    def save
      self.class.ensure_api_key_set
      raise InvalidParameterException.new("name is required") if self.class.blank?(@name)
      raise InvalidParameterException.new("crontab is required") if self.class.blank?(@crontab)

      attributes = {
        :name => @name,
        :crontab => @crontab,
        :time_zone => @time_zone,
        :machine => @machine,
        :finish_alarms_enabled => @finish_alarms_enabled,
        :maximum_run_time => @maximum_run_time,
        :start_notification_grace_period => @start_notification_grace_period,
        :consecutive_alarmed_tasks_required_to_trigger_alarm => @consecutive_alarmed_tasks_required_to_trigger_alarm
      }

      Proby.logger.info "Updating task #{@api_id} with attributes: #{attributes.inspect}"
      response = self.class.put("/api/v1/tasks/#{@api_id}.json",
                                :format => :json,
                                :body => MultiJson.encode(:task => attributes),
                                :headers => self.class.default_headers)

      if response.code == 200
        true
      else
        self.class.handle_api_failure(response)
      end 
    end

    # Delete a Proby task.  The object will be frozen after the delete.
    #
    # @example
    #   proby_task = ProbyTask.get('my_proby_task_api_id')
    #   proby_task.delete
    def delete
      self.class.ensure_api_key_set

      Proby.logger.info "Deleting task #{@api_id}"
      response = self.class.delete("/api/v1/tasks/#{@api_id}.json",
                                   :format => :json,
                                   :headers => self.class.default_headers)

      if response.code == 200 
        self.freeze
        true
      else
        self.class.handle_api_failure(response)
      end 
    end

    # Pause a Proby task.
    #
    # @example
    #   proby_task = ProbyTask.get('my_proby_task_api_id')
    #   proby_task.pause
    def pause
      self.class.ensure_api_key_set

      Proby.logger.info "Pausing task #{@api_id}"
      response = self.class.post("/api/v1/tasks/#{@api_id}/pause.json",
                                 :format => :json,
                                 :headers => self.class.default_headers)

      if response.code == 200 
        @paused = true
        true
      else
        self.class.handle_api_failure(response)
      end 
    end

    # Unpause a Proby task.
    #
    # @example
    #   proby_task = ProbyTask.get('my_proby_task_api_id')
    #   proby_task.unpause
    def unpause
      self.class.ensure_api_key_set

      Proby.logger.info "Unpausing task #{@api_id}"
      response = self.class.post("/api/v1/tasks/#{@api_id}/unpause.json",
                                 :format => :json,
                                 :headers => self.class.default_headers)

      if response.code == 200 
        @paused = false
        true
      else
        self.class.handle_api_failure(response)
      end 
    end

    private

    def self.list
      Proby.logger.info "Getting the list of tasks"
      response = get('/api/v1/tasks.json',
                     :format => :json,
                     :headers => default_headers)

      if response.code == 200
        data = response.parsed_response['tasks']
        data.map { |task_data| new(task_data) }
      else
        handle_api_failure(response)
      end
    end

    def self.fetch(api_id)
      raise InvalidParameterException.new("api_id is required") if api_id.nil? || api_id.strip.empty?

      Proby.logger.info "Fetching task from Proby: #{api_id}"
      response = get("/api/v1/tasks/#{api_id}.json",
                     :format => :json,
                     :headers => default_headers)

      if response.code == 200
        new(response.parsed_response['task'])
      elsif response.code == 404
        nil
      else
        handle_api_failure(response)
      end
    end

    def self.ensure_api_key_set
      if Proby.api_key.nil? || Proby.api_key.strip.empty?
        raise InvalidApiKeyException.new("Your Proby API key has not been set.  Set it using Proby.api_key = 'my_api_key'")
      end
    end

  end
end
